migrate hotornot user tokens & posts to yral

rosarp-gobazzinga opened the pull request.

rosarp committed:
migrate hotornot user tokens & posts to yral

rosarp-gobazzinga committed:
renamed mod

rosarp-gobazzinga commented:
#### Completed:

- [ ] transfer token from hotornot id
- [ ] receive token in yral id
- [ ] copy posts to yral id
- [ ] check: hotornot id belongs to HOT_OR_NOT_SUBNET
- [ ] check: yral id does not belongs to HOT_OR_NOT_SUBNET

#### TODO:

- [ ] Tests fix
- [ ] Error: individual_user_template.wasm.gz missing
- [ ] Session change for hotornot user -> MigrateSession
- [ ] New tests??
- [ ] candid generation
- [ ] Any more checks required??

rosarp-gobazzinga committed:
fixing build error

rosarp-gobazzinga committed:
updating can.did for individual_user_template

rosarp-gobazzinga committed:
added MigrationStatus flag for user
only 1 a/c can be migrated to yral a/c

rosarp-gobazzinga committed:
added migration_status to Snapshot for backups

komal-rs commented on src/canister/individual_user_template/src/api/canister_management/migrate_hotornot_user_to_yral.rs:
```diff
@@ -0,0 +1,158 @@
+
+    // Should be in hotornot subnet
+    if check_canister_is_in_hotornot_subnet(profile_owner, true)
+        .await
```
Shoud we also check if it result is Ok(false)

rosarp-gobazzinga replied:
if Ok(false) condition, which means, it is not part of hotornot subnet. In this case we want to continue. So no need to check this as we are continueing the flow.

rosarp-gobazzinga replied:
Realized the conditions to check is false here. And for true, the flow continues. Will fix it.

komal-rs commented on src/canister/individual_user_template/src/api/canister_management/migrate_hotornot_user_to_yral.rs:
```diff
@@ -0,0 +1,158 @@
+
+    // Should not be in hotornot subnet
+    if check_canister_is_in_hotornot_subnet(profile_owner, false)
+        .await
```
Same issue as above

rosarp-gobazzinga replied:
Realized the conditions to check is true here. And for false, the flow continues, so no need to have else block. Will fix it.

komal-sai-yral reviewed (changes requested):
Pls add a pocket-ic integration test.

rosarp-gobazzinga committed:
fixed condition to check subnet
updated comments

rosarp-gobazzinga commented:
New changes TODO:

- yral principal id should be authenticated otherwise transfer should be deined

ravibazz commented on src/lib/shared_utils/src/common/types/utility_token/token_event.rs:
```diff
@@ -15,7 +15,16 @@ pub enum TokenEvent {
     },
     Burn,
-    Transfer,
+    Transfer {
+        amount: u64,
+        to_account: Principal,
+        timestamp: SystemTime,
+    },
+    Receive {
+        amount: u64,
+        from_account: Principal,
+        timestamp: SystemTime,
+    },
```
I think lets use more descriptive name here. `HotorNotTransfer` and `HotorNotRecieve`

ravibazz commented on src/canister/individual_user_template/src/lib.rs:
```diff
@@ -36,6 +36,7 @@ use shared_utils::{
         GetUserUtilityTokenTransactionHistoryError, UpdateProfileSetUniqueUsernameError,
     },
 };
+use std::collections::BTreeMap;
```
Why do we need this?

rosarp-gobazzinga replied:
[migrate_hotornot_user_to_yral.rs](https://github.com/go-bazzinga/hot-or-not-backend-canister/pull/303/files/4d4379108167cd71bd6e422f2663b29d30ae1cc6#diff-b05e5b6b862550c125daf5a61aeb25173d297a5ec5641a83f44cd2cd39416e90) needs BTreeMap. Is there any other file this can be moved? use in migrate_hotornot_user_to_yral file alone doesnt work. export_candid fails.

ravibazz commented on src/canister/individual_user_template/src/api/canister_management/migrate_hotornot_user_to_yral.rs:
```diff
@@ -0,0 +1,158 @@
+const HOT_OR_NOT_CONTROLLER: &str = "rimrc-piaaa-aaaao-aaljq-cai";
+
+#[update]
+pub async fn transfer_tokens_and_posts(to_account: Principal) -> Result<String, String> {
```
Need to mark migrated for this canister as well. So hotornot canister shouldn't be accessible after migration.

rosarp-gobazzinga replied:
i'm making enum with 3 values. not started, migrated_from_hotornot { principal_id }, migrated_to_yral { principal_id }
So from UI, if this value is set, canister is already migrated.

ravibazz commented on src/canister/individual_user_template/src/api/canister_management/migrate_hotornot_user_to_yral.rs:
```diff
@@ -0,0 +1,158 @@
+        return Err("Already Migrated".to_owned());
+    };
+
+    let profile_owner =
+        CANISTER_DATA.with_borrow(|canister_data| canister_data.profile.principal_id.unwrap());
+
+    if profile_owner != caller() {
+        return Err("Unauthorized".to_owned());
+    }
```
This wouldn't be called from profile owner. This will be called from hotornot canister profile owner. We can only check if that canister belongs to hotornot subnet.

rosarp-gobazzinga committed:
adding get_profile_details_v2
adds migration_info in v2
adding test cases for migrate user functionality

rosarp-gobazzinga committed:
Mocked hot-or-not canister id for tests
Added migrate_token test (WIP)

rosarp-gobazzinga committed:
ic_cdk call to have canister principal id
using user_index to find to_account's canister id
to call receive method for to_account principal id

rosarp-gobazzinga committed:
updated ic_cdk::call issue fixed

rosarp-gobazzinga committed:
moved fetching canister id to off-chain-agent

rosarp-gobazzinga committed:
fixing assert in test

rosarp-gobazzinga committed:
Merge remote-tracking branch 'refs/remotes/origin/main'

rosarp-gobazzinga committed:
Merge branch 'main' into feat/transfer-hotornot-tokens-to-yral

rosarp-gobazzinga committed:
updated test with change of params for transfer fn

Diksha894 commented:
### Menu Changes

 - [ ] Add “HotorNot Account Transfer” option in Menu - Below “nsfw” and above “refer and earn”
    
    Follow this screen: https://www.figma.com/design/ig9dxeJrajuzHNGlo4zUps/experiment-file?node-id=1054-3238&t=ExUCGcNaobfWnjFl-4
    
 “HotorNot Account Transfer” screen
 - [ ] Not Logged in version: https://www.figma.com/design/ig9dxeJrajuzHNGlo4zUps/experiment-file?node-id=1055-3483&t=ExUCGcNaobfWnjFl-4
        
        This has a LOGIN button and no principal ID is shown
        
 - [ ] Logged in version: https://www.figma.com/design/ig9dxeJrajuzHNGlo4zUps/experiment-file?node-id=1099-3321&t=ExUCGcNaobfWnjFl-4
        
        This has the principal ID to be copied - with an “ID copied” feedback if a user clicks on “copy” button
        
         It also has a hyperlink back to hotornot (link to be provided later)
        
 - [ ] ON DESKTOP MENU: Truncate the principal ID when user is not logged in https://www.figma.com/design/ig9dxeJrajuzHNGlo4zUps/experiment-file?node-id=567-10463&t=ExUCGcNaobfWnjFl-4

### Wallet Changes

 - [ ] ON DESKTOP WALLET: Truncate the principal ID when user is not logged in https://www.figma.com/design/ig9dxeJrajuzHNGlo4zUps/experiment-file?node-id=567-8245&t=ExUCGcNaobfWnjFl-4

### Profile Page Changes

Changes to be made to “before login state” of profile on both desktop and phone versions https://www.figma.com/design/ig9dxeJrajuzHNGlo4zUps/experiment-file?node-id=567-10156&t=ExUCGcNaobfWnjFl-4

 - [ ] Remove the **@username** completely
 - [ ] Show only one username (which is the principal ID) - Truncated 
 - [ ] Earnings, Login button and login text to be below the profile picture and username

*optional change to be done if implementation is very similar to the above change

Changes to be made to “after login” state of profile on both desktop and phone versions https://www.figma.com/design/ig9dxeJrajuzHNGlo4zUps/experiment-file?node-id=987-3320&t=ExUCGcNaobfWnjFl-4

 - [ ] Remove the @username completely
 - [ ] Show only one username - which is the principal ID (not truncated here)
 - [ ] Earnings below the profile picture and username

Commit:
move migration logic to individual user trait

Commit:
extend global know principal from platform orchestrator while provisioning new subnet orchestrators

Commit:
fix tests

Commit:
added additional tests

Commit:
fix a test

Commit:
add serde default for new field

Commit:
Merge branch 'main' into feat/transfer-hotornot-tokens-to-yral

Commit:
fix ic-repl script

Commit:
added test for post transfer

Commit:
remove mockall dependency

komal-rs commented on src/canister/individual_user_template/src/util/migration.rs:
```diff
@@ -0,0 +1,203 @@
+            CANISTER_DATA.with_borrow(|canister_data| {
+                canister_data
+                    .known_principal_ids
+                    .get(&KnownPrincipalType::CanisterIdHotOrNotSubnetOrchestrator)
```
Are we sure KnownPrincipalType::CanisterIdHotOrNotSubnetOrchestrator is present in all individual canisters ?

ravibazz replied:
No we will use platform-orchestrator to issue global id that will reflect in all canisters

komal-rs replied:
which method is this ?

komal-rs commented on src/canister/individual_user_template/src/api/canister_management/migrate_hotornot_user_to_yral.rs:
```diff
@@ -0,0 +1,34 @@
+}
+
+#[update]
+pub async fn receive_data_from_hotornot(
```
can a bad actor canister call this and get as many token as it needs ?

komal-rs replied:
I think one simple check would be to call the subnet orchestrator and see if the canister is present in the subnet it claims

ravibazz replied:
No it cannot happen as `IndividualUser::new` calls CanisterInfo. If a user calls this method  `IndividualUser::new` will fail as it will be a call from a canister.

komal-rs replied:
No ignore. I think one check at the user.recieve_tokens_and_posts saying if the canister subnet type is HotorNot works

ravibazz replied:
but I will add a test for this. Good catch 😄

ravibazz replied:
Added a check in `recieve` method to ensure the caller is from HotorNot subnet and added a test for the same use case.

Commit:
handle the case where user tries to misuse the functionality by creating dummy canister
